swimlane/core/client.py

```diff
@@ -248,52 +250,67 @@ def user(self):
 class SwimlaneAuth(SwimlaneResolver):
     """Handles authentication for all requests"""
 
+    _token_expiration_buffer = pendulum.Interval(minutes=5)
+
     def __init__(self, swimlane, username, password, verify_ssl=True):
         super(SwimlaneAuth, self).__init__(swimlane)
 
-        self.user, self._login_headers = self.authenticate(username, password, verify_ssl)
+        self._username = username
+        self._password = password
+        self._verify_ssl = verify_ssl
+
+        self.user = None
+        self._login_headers = {}
+        self._token_expiration = pendulum.now()
 
     def __call__(self, request):
+        """Attach necessary headers to all requests
+
+        Automatically reauthenticate before sending request when nearing token expiration
+        """
+
+        # Refresh token if it expires soon
+        if pendulum.now() + self._token_expiration_buffer >= self._token_expiration:
+            self.authenticate()
 
         request.headers.update(self._login_headers)
 
         return request
 
-    def authenticate(self, username, password, verify_ssl=True):
-        """Send login request and return User instance and login headers"""
-        # Explicitly provide verify_ssl argument, appears to not consistently be acknowledged across versions during
-        # initial setup for auth
+    def authenticate(self):
+        """Send login request and update User instance, login headers, and token expiration"""
+
+        # Temporarily remove auth from Swimlane session for auth request to avoid recursive loop during login request
+        self._swimlane._session.auth = None
         resp = self._swimlane.request(
             'post',
             'user/login',
             json={
-                'userName': username,
-                'password': password,
+                'userName': self._username,
+                'password': self._password,
                 'domain': ''
```

Comment:
Should be able to get away without supplying `domain` as part of the JSON, it isn't used by Swimlane

Reply:
Believe that's a holdover from pre-2.14 days, missed that taking out the rest of that support, will remove.